File: frontend/src/types/packing.ts
/** Contract v1: docs/CONTRACTS.md. Integer millimetres and grams; volumes in mm³. */
export type Orientation = 'LWH' | 'LHW' | 'WLH' | 'WHL' | 'HLW' | 'HWL';
export type PackingStatus = 'success' | 'partial' | 'impossible';
export type IssueSeverity = 'info' | 'warning' | 'error';
export type IssueCode =
  | 'ITEM_TOO_LARGE'
  | 'ITEM_TOO_HEAVY'
  | 'BOX_STOCK_EXHAUSTED'
  | 'NO_BOX_TYPES'
  | 'NO_FEASIBLE_PLACEMENT'
  | 'PARTIAL_PACKING'
  | 'SIMILAR_ALTERNATIVES'
  | 'DEMO_STUB';

export interface Dimensions {
  length: number;
  width: number;
  height: number;
}

export interface BoxType extends Dimensions {
  id: string;
  name: string;
  max_weight: number;
  available_count: number;
}

export interface Product extends Dimensions {
  id: string;
  name: string;
  weight: number;
  quantity: number;
  allow_rotation: boolean;
}

export interface ItemInstance extends Dimensions {
  id: string;
  product_id: string;
  name: string;
  unit_index: number;
  weight: number;
  allow_rotation: boolean;
}

/** Minimum corner in the box: x right, y toward the back, z up. */
export interface Position {
  x: number;
  y: number;
  z: number;
}

export interface Placement {
  item_instance_id: string;
  product_id: string;
  position: Position;
  dimensions: Dimensions;
  orientation: Orientation;
  step: number;
}

export interface PackingInstructionStep {
  step: number;
  action: 'prepare_box' | 'place_item' | 'close_box';
  box_id: string;
  message: string;
  item_instance_id: string | null;
  product_id: string | null;
  position: Position | null;
  dimensions: Dimensions | null;
  orientation: Orientation | null;
}

export interface PackedBox extends Dimensions {
  id: string;
  box_type_id: string;
  name: string;
  max_weight: number;
  total_weight: number;
  used_volume: number;
  fill_ratio: number;
  placements: Placement[];
  instructions: PackingInstructionStep[];
}

export interface PackingMetrics {
  total_items: number;
  packed_items: number;
  unpacked_items: number;
  boxes_used: number;
  boxes_by_type: Record<string, number>;
  total_box_volume: number;
  used_volume: number;
  empty_volume: number;
  fill_ratio: number;
  total_weight: number;
}

export interface PackingIssue {
  code: IssueCode;
  severity: IssueSeverity;
  message: string;
  item_instance_ids: string[];
  box_type_ids: string[];
}

export interface PackingAlternative {
  id: string;
  description: string;
  status: PackingStatus;
  metrics: PackingMetrics;
  packed_boxes: PackedBox[];
  unpacked_items: ItemInstance[];
  issues: PackingIssue[];
}

export interface PackingOptions {
  include_alternatives?: boolean;
  max_alternatives?: number;
  algorithm?: PackingAlgorithm;
  solver_timeout_ms?: number;
  solver_workers?: number;
}

export type PackingAlgorithm = 'heuristic' | 'z3';

export interface PackingOptimization {
  status: 'optimal' | 'feasible' | 'fallback';
  reason: 'completed' | 'time_limit' | 'size_limit' | 'solver_error';
  workers: number;
  time_limit_ms: number;
  support_ratio: number;
}

export interface PackingRequest {
  boxes: BoxType[];
  products: Product[];
  options?: PackingOptions;
}

export interface PackingResult {
  status: PackingStatus;
  metrics: PackingMetrics;
  packed_boxes: PackedBox[];
  unpacked_items: ItemInstance[];
  issues: PackingIssue[];
  alternatives: PackingAlternative[];
  algorithm_version: string;
  optimization?: PackingOptimization | null;
  calculation_seconds?: number | null;
}

export interface DemoScenario {
  id: string;
  name: string;
  description: string;
  expected_status: PackingStatus;
}

export interface HealthResponse {
  status: 'ok';
  api_version: 'v1';
  engine: string;
}

export type ApiErrorCode =
  | 'HTTP_ERROR'
  | 'METHOD_NOT_ALLOWED'
  | 'VALIDATION_ERROR'
  | 'NOT_FOUND'
  | 'CONFLICT'
  | 'ENGINE_NOT_IMPLEMENTED'
  | 'INTERNAL_ERROR';

export interface ApiErrorDetail {
  field: string;
  message: string;
  type: string;
}

export interface ApiErrorResponse {
  error: {
    code: ApiErrorCode;
    message: string;
    details: ApiErrorDetail[];
  };
}

File: backend/app/schemas/packing.py
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, StringConstraints, model_validator

from app.domain import models as domain

Identifier = Annotated[str, StringConstraints(pattern=r"^[A-Za-z0-9][A-Za-z0-9_-]{0,63}$")]
Name = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=200)]
Dimension = Annotated[int, Field(strict=True, gt=0, le=100_000)]
Weight = Annotated[int, Field(strict=True, gt=0, le=100_000_000)]
NonNegativeInt = Annotated[int, Field(strict=True, ge=0)]
PositiveInt = Annotated[int, Field(strict=True, gt=0)]
Ratio = Annotated[float, Field(ge=0, le=1, allow_inf_nan=False)]


class ContractModel(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)


class BoxTypeSchema(ContractModel):
    id: Identifier
    name: Name
    length: Dimension
    width: Dimension
    height: Dimension
    max_weight: Weight
    available_count: NonNegativeInt

    def to_domain(self) -> domain.BoxType:
        return domain.BoxType(**self.model_dump())


class ProductSchema(ContractModel):
    id: Identifier
    name: Name
    length: Dimension
    width: Dimension
    height: Dimension
    weight: Weight
    quantity: PositiveInt
    allow_rotation: Annotated[bool, Field(strict=True)] = True

    def to_domain(self) -> domain.Product:
        return domain.Product(**self.model_dump())


class PackingOptionsSchema(ContractModel):
    include_alternatives: Annotated[bool, Field(strict=True)] = True
    max_alternatives: Annotated[int, Field(strict=True, ge=0, le=5)] = 3
    algorithm: domain.PackingAlgorithm = "heuristic"
    solver_timeout_ms: PositiveInt = 10_000
    solver_workers: PositiveInt = 4


class PackingRequestSchema(ContractModel):
    boxes: list[BoxTypeSchema]
    products: Annotated[list[ProductSchema], Field(min_length=1)]
    options: PackingOptionsSchema = Field(default_factory=PackingOptionsSchema)

    @model_validator(mode="after")
    def validate_order(self) -> "PackingRequestSchema":
        for name, entries in (("boxes", self.boxes), ("products", self.products)):
            ids = [entry.id for entry in entries]
            if len(ids) != len(set(ids)):
                raise ValueError(f"{name}: duplicate id")
        return self

    def to_domain(self) -> domain.PackingRequest:
        return domain.PackingRequest(
            boxes=tuple(box.to_domain() for box in sorted(self.boxes, key=lambda box: box.id)),
            products=tuple(
                product.to_domain()
                for product in sorted(self.products, key=lambda product: product.id)
            ),
            options=domain.PackingOptions(**self.options.model_dump()),
        )


class PositionSchema(ContractModel):
    x: NonNegativeInt
    y: NonNegativeInt
    z: NonNegativeInt


class DimensionsSchema(ContractModel):
    length: Dimension
    width: Dimension
    height: Dimension


class ItemInstanceSchema(ContractModel):
    id: str
    product_id: Identifier
    name: Name
    unit_index: PositiveInt
    length: Dimension
    width: Dimension
    height: Dimension
    weight: Weight
    allow_rotation: bool


class PlacementSchema(ContractModel):
    item_instance_id: str
    product_id: Identifier
    position: PositionSchema
    dimensions: DimensionsSchema
    orientation: domain.Orientation
    step: PositiveInt


class PackingInstructionStepSchema(ContractModel):
    step: NonNegativeInt
    action: Literal["prepare_box", "place_item", "close_box"]
    box_id: str
    message: str
    item_instance_id: str | None = None
    product_id: Identifier | None = None
    position: PositionSchema | None = None
    dimensions: DimensionsSchema | None = None
    orientation: domain.Orientation | None = None


class PackedBoxSchema(ContractModel):
    id: str
    box_type_id: Identifier
    name: Name
    length: Dimension
    width: Dimension
    height: Dimension
    max_weight: Weight
    total_weight: NonNegativeInt
    used_volume: NonNegativeInt
    fill_ratio: Ratio
    placements: list[PlacementSchema]
    instructions: list[PackingInstructionStepSchema]


class PackingMetricsSchema(ContractModel):
    total_items: NonNegativeInt
    packed_items: NonNegativeInt
    unpacked_items: NonNegativeInt
    boxes_used: NonNegativeInt
    boxes_by_type: dict[str, NonNegativeInt]
    total_box_volume: NonNegativeInt
    used_volume: NonNegativeInt
    empty_volume: NonNegativeInt
    fill_ratio: Ratio
    total_weight: NonNegativeInt


class PackingIssueSchema(ContractModel):
    code: domain.IssueCode
    severity: Literal["info", "warning", "error"]
    message: str
    item_instance_ids: list[str]
    box_type_ids: list[Identifier]


class PackingAlternativeSchema(ContractModel):
    id: str
    description: str
    status: domain.PackingStatus
    metrics: PackingMetricsSchema
    packed_boxes: list[PackedBoxSchema]
    unpacked_items: list[ItemInstanceSchema]
    issues: list[PackingIssueSchema]


class OptimizationInfoSchema(ContractModel):
    status: Literal["optimal", "feasible", "fallback"]
    reason: Literal["completed", "time_limit", "size_limit", "solver_error"]
    workers: NonNegativeInt
    time_limit_ms: PositiveInt
    support_ratio: Ratio = 1.0


class PackingResultSchema(ContractModel):
    status: domain.PackingStatus
    metrics: PackingMetricsSchema
    packed_boxes: list[PackedBoxSchema]
    unpacked_items: list[ItemInstanceSchema]
    issues: list[PackingIssueSchema]
    alternatives: list[PackingAlternativeSchema]
    algorithm_version: str
    optimization: OptimizationInfoSchema | None = None
    calculation_seconds: float | None = Field(default=None, ge=0)
